Packaging: Add venv and packs to deb/rpm packages

cognifloyd opened the pull request:
_Reviewers: Please respond to the **Open Questions** section below._

This PR is working towards doing packaging via pantsbuild. Eventually, I hope to archive and stop using st2-packages.git.

Previous PRs in this series:
- The pants nfpm backend was activated in #6321.
- Basic BUILD metadata for deb/rpm packages (`nfpm_deb_package` and `nfpm_rpm_package`) was added in #6325 along with the deb maintainer scripts and rpm scriptlets.
- #6326 added the systemd files (units and generators) to the deb/rpm packages
- #6327 added directories and symlinks to the deb/rpm packages

This PR adds the virtualenv (by way of `st2.pex` added in #6307) and packs (by way of pack archives added in #6311). Both st2.pex and the pack archives get extracted by the post install deb maintainer script or rpm scriptlet (added in #6320).

### Virtualenv (`st2.pex`)

For the virtualenv, we add an `nfpm_content_file` target for `st2.pex`:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.venv#L84-L90

This target is parametrized to produce a pex for each python minor version we support. Because the parametrization was very repetitive, I added this helper to simplify defining each of them:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.venv#L76-L81

Then, this is how we use that helper to actually parametrize the st2.pex target for each python minor version:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.venv#L91-L94

Then we add a dependency to the `nfpm_content_file` target on the `nfpm_deb_package` target and the `nfpm_rpm_package` target:
https://github.com/StackStorm/st2/blob/1ebd1995f69a197eaceca78cb2251fd623958f32/packaging/BUILD#L25-L31
https://github.com/StackStorm/st2/blob/1ebd1995f69a197eaceca78cb2251fd623958f32/packaging/BUILD#L50
https://github.com/StackStorm/st2/blob/1ebd1995f69a197eaceca78cb2251fd623958f32/packaging/BUILD#L85

I have an open question (below) about how tightly we should couple a python version to an OS. Based on feedback here and in the TSC meeting, we will not couple the OS with the python version. If someone wants to use pyenv or otherwise install a different (typically newer) version of python, they should be able to do that simply without fighting with the system package.

### Packs (pack archives)

First, we add an `nfpm_content_file` target to the `st2_pack_archive(...)` macro:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/pants-plugins/macros.py#L173-L181

Then we add some generic `target`s to make it easier for the deb/rpm packages to depend on all the pack archives.
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.packs#L17-L25

The post-install scripts only install the default packs, which are defined in 3 places:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.packs#L1-L8
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/deb/scripts/post-install.sh#L32-L39
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/rpm/scripts/post-install.sh#L10-L17

Plus the post-install scripts install the `examples` pack in `/usr/share/doc/st2/examples`:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/deb/scripts/post-install.sh#L113
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/rpm/scripts/post-install.sh#L49

So, we need the packs that get installed by default, and the `examples` pack. The rest of the packs are optional, and have not been made available in older versions of st2. There is an open question (below) about these extra packs:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD.packs#L10-L15

Finally, we add a dependency on the `nfpm_content_file` targets to the `nfpm_deb_package` and `nfpm_rpm_package` targets:
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD#L42-L43
https://github.com/StackStorm/st2/blob/850cc78c25c2a90fa3cc77dde159358cb95f8d93/packaging/BUILD#L77-L78

### Open Questions

1. How tightly should we couple a python version to an OS?
    1. ~We could include only one `st2.pex` on each OS based on which python version is packaged on that OS.~
        - In this scenario, the pex gets installed as `/opt/stackstorm/install/st2.pex`.
        - This is what I've done by using the dependency: `:st2_venv@parametrize=py38`.
        - This is also the most efficient use of disk space, as only one copy of the pex gets installed, not 4.
    2. :heavy_check_mark: We could include an `st2.pex` file for each python version we support, on every OS, so that people can easily rebuild the venv when they update to a newer python version.
        - :heavy_check_mark: In this scenario, we need to modify the `_venv_py3(minor)` helper function to define `dst=f"/opt/stackstorm/install/st2-py3{minor}.pex"`
        - In this scenario, we also need to either:
            1. ~modify the post-install script to pick which pex to use, or~
            2. :heavy_check_mark: add a symlink for the default python version as `/opt/stackstorm/install/st2.pex -> /opt/stackstorm/install/st2-py38.pex`. Maybe we can record this symlink as a `config|noreplace` file so that people can modify their preferred python version.
        - This would allow people to update to newer python versions using something like `pyenv` or a `ppa` to install a newer version of python than the OS packages provide.
2. Should we install the pack archives for all of the optional extra packs?
    - The `examples` pack is not optional, as we install it in `/usr/share/doc/st2/examples`.
    - Making the `debug` pack easier to install would be great, so I'd like to include its pack archive.
    - Should we include the `hello_st2` pack archive? This is the "answer key" for the ["Creating your First Pack" tutorial](https://docs.stackstorm.com/reference/packs.html?highlight=hello_st2#creating-your-first-pack).

guzzijones commented:
The Enterprise Linux versions are lagging behind python.  I vote:
>We could include an st2.pex file for each python version we support, on every OS, so that people can easily rebuild the venv when they update to a newer python version.

    In this scenario, we need to modify the _venv_py3(minor) helper function to define dst=f"/opt/stackstorm/install/st2-py3{minor}.pex"
        add a symlink for the default python version as /opt/stackstorm/install/st2.pex -> /opt/stackstorm/install/st2-py38.pex. Maybe we can record this symlink as a config|noreplace file so that people can modify their preferred python version.

cognifloyd commented:
I modified the package deps so that pexes for all supported python versions are included. I also modified the post-install.sh script to create a symlink at `/opt/stackstorm/install/st2.pex` if the symlink is missing or is broken.

I included a bugfix for `pants-plugins/release` (I forgot to set some dataclasses as frozen).
I also included a modification to the compression algorithm used for building rpms. `xz` might be the default for el8, but it's significantly slower at over 3 minutes to build the rpm vs <1 minute for `zstd`. So, now we just use `zstd:best` compression for rpms.

NOTE: when we start making releases with pants, we need to:
- announce in the release notes that all deb/rpm package installs/upgrades will **remove and rebuild** the `/opt/stackstorm/st2` virtualenv. So, any modifications to the venv will need to be re-applied after upgrades.
- add instructions to the docs on rebuilding the st2 venv (update the st2.pex symlink and run it) to upgrade python.
